docs/proposals/20230117-cluster-level-identity.md

```diff
@@ -0,0 +1,166 @@
+individual workload clusters has to be created using OCI user principals tied to individual workload clusters.
+
+### Custom Resource Changes
+
+The following fields will be added to OCIClusterSpec custom resource.
+
+```go
+// OCIClusterSpec defines the desired state of OciCluster
+type OCIClusterSpec struct {
+  // IdentityRef is a reference to an identity(principal) to be used when reconciling this cluster
+  // +optional
+  IdentityRef *corev1.ObjectReference `json:"identityRef,omitempty"`
+}
+```
+The following custom resources will be added.
+
+```go
+
+type PrincipalType string
+
+const (
+    // UserPrincipal represents a user princpal.
+    UserPrincipal PrincipalType = "UserPrincipal"
+)
+
+// OCIClusterIdentitySpec defines the parameters that are used to create an OCIIdentity.
+type OCIClusterIdentitySpec struct {
+  // Type is the type of OCI Principal used.
+  // UserPrincipal is the only supported value
+  Type PrincipalType `json:"type"`
+  
+  // PrincipalSecret is a secret reference which contains the authentication credentials for the principal.
+  // +optional
+  PrincipalSecret corev1.SecretReference `json:"clientSecret,omitempty"`
+
+  // AllowedNamespaces is used to identify the namespaces the clusters are allowed to use the identity from.
+  // Namespaces can be selected either using an array of namespaces or with label selector.
+  // An empty allowedNamespaces object indicates that OCIClusters can use this identity from any namespace.
+  // If this object is nil, no namespaces will be allowed (default behaviour, if this field is not provided)
+  // A namespace should be either in the NamespaceList or match with Selector to use the identity.
+  //
+  // +optional
+  // +nullable
+  AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+// OCIClusterIdentityStatus defines the observed state of OCIClusterIdentity.
+type OCIClusterIdentityStatus struct {
+  // Conditions defines current service state of the OCIClusterIdentity.
+  // +optional
+  Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+type OCIClusterIdentity struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec   OCIClusterIdentitySpec   `json:"spec,omitempty"`
+  Status OCIClusterIdentityStatus `json:"status,omitempty"`
+}
+
+```
+
+
+### Example Cluster template
+
+```yaml
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OCICluster
+spec:
+  identityRef:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+    kind: OCIClusterIdentity
+    name: <test-identity>
+    namespace: <test>
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OCIClusterIdentity
+metadata:
+  name: test-identity
+  namespace: test
+spec:
+  type: UserPrincipal
+  clientSecret: 
+    name: user-credentials
+    namespace: secret-namespace
+  allowedNamespaces:
+    list:
+      - test
+---
+apiVersion: v1
```

Comment:
What would be the behaviour when secrets are modified or updated. Will the controllers referencing them update automatically at their end as well. ?

Reply:
So in the initial implementation, I am thinking, each reconcile, we will read the secret and create the clients, since that should not be too expensive anbd that is what other providers are doing. So when a secret is updated there are 2 cases
1) if the secret was already wrong and the reconcile was erroring out, the next reconcile based on the exponential backoff will use the new secret
2) If the secret was good,  the next reconcile , which happens every five mins will use it.